WcaOnRails/spec/jobs/registration_reminder_job_spec.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe RegistrationReminderJob, type: :job do
+  describe "registration reminder job" do
+    let(:user) { FactoryBot.create :user }
+    let(:delegate) { FactoryBot.create :delegate }
+    let(:organizers) { FactoryBot.create_list :user, 3 }
+    let(:competition) { FactoryBot.create :competition, :visible, organizers: organizers, delegates: [delegate] }
+
+    it "does not send more than 24h in advance" do
+      competition.registration_open = 2.days.after
+
+      expect do
+        RegistrationReminderJob.perform_now
+      end.to change { ActionMailer::Base.deliveries.length }.by(0)
+    end
+
+    it "schedules registration reminder emails" do
+      BookmarkedCompetition.create(competition: competition, user: user)
+      competition.update_column(:registration_open, 12.hours.after)
+
+      expect(CompetitionsMailer).to receive(:registration_reminder).with(competition, user).and_call_original
+      expect(CompetitionsMailer).to receive(:registration_reminder).with(competition, delegate).and_call_original
+      organizers.each do |organizer|
+        expect(CompetitionsMailer).to receive(:registration_reminder).with(competition, organizer).and_call_original
+      end
+
+      expect do
+        RegistrationReminderJob.perform_now
+      end.to change { ActionMailer::Base.deliveries.length }.by(5)
+
+      # Running the job again won't send any more emails.
+      expect do
+        RegistrationReminderJob.perform_now
+      end.to change { ActionMailer::Base.deliveries.length }.by(0)
+    end
+
+    it "resends when registration period changes" do
+      BookmarkedCompetition.create(competition: competition, user: user)
+      competition.update_column(:registration_reminder_sent_at, 7.days.ago)
+      competition.update_column(:registration_open, 12.hours.after)
+
+      expect(CompetitionsMailer).to receive(:registration_reminder).with(competition, user).and_call_original
+      expect(CompetitionsMailer).to receive(:registration_reminder).with(competition, delegate).and_call_original
+      organizers.each do |organizer|
+        expect(CompetitionsMailer).to receive(:registration_reminder).with(competition, organizer).and_call_original
+      end
+
+      expect do
+        RegistrationReminderJob.perform_now
+      end.to change { ActionMailer::Base.deliveries.length }.by(5)
+    end
+
+    it "does not send to registered and accepted users" do
```

Comment:
Should it matter if the registration is accepted or not?

Reply:
Aha, then maybe we should add something to the email to make it clear that we know the user is registered, but its registration hasn't been accepted yet.

Reply:
Done.